2a_model/src/do_it_functions.py

```diff
@@ -0,0 +1,183 @@
+           'site_min_confidence', 'site_name', 'sph', 'srad', 'temp.water', 'tmmn',
+           'tmmx', 'velocity', 'vs']]
+    targets_site = inputs_df.loc[site][['do_min','do_mean','do_max']]
+    
+    if sink == 'do_range':
+        targets_site['do_range'] = targets_site['do_max']-targets_site['do_min']
+        
+    tar_dict = {}
+
+    model_preds = pd.read_feather(predictions_file)
+    model_preds = model_preds[model_preds['site_id'] == site].set_index('date')[['do_min','do_mean','do_max']]
+    model_preds['do_range'] = model_preds['do_max']-model_preds['do_min']
+    #create targets dictionary
+    if model != 'observed':
+        tar_dict[model] = model_preds
+    tar_dict['observed'] = targets_site
+
+
+    
+    #create dictionary to store calculations in
+    max_it = {}
+    #create a nested dictionary for each DO variable to store it calcs
+    #TE0 = Transfer Entropy at a time lag of 0, MI = mututal information,
+    #TEmax is the maximum TE, TEmaxT is the time lag of the maximum TE,
+    #TEmaxcrit is the a True/False if TEmax is significant, everything follows the same 
+    #convention for MI
+    
+    
+    #join input and target to make sure they are aligned
+    site_inptar = inputs_site.join(tar_dict[model], rsuffix = '_pred')
+    
+    #assign x = source, y = sink
+    x = site_inptar[source]
+    y = site_inptar[sink]
+    
+    #for calculating rmse
+    obs_pred = tar_dict['observed'][[sink]].join(tar_dict[model][[sink]], rsuffix = '_pred')
+    
+    
+    #load the preprocessing fucntions from it_functions.py
+    ppf = it_functions.pre_proc_func()
+    
+    if log_transform:
+        xl10 = ppf.log10(x)
+    else:
+        xl10 = x.copy()
+    x_rss = ppf.remove_seasonal_signal(xl10)
+    x_ss = ppf.standardize(x_rss)
+    
+    y_prepped = {}
+    y_rss = ppf.remove_seasonal_signal(y)
+    y_prepped = ppf.standardize(y_rss)
+    
+    
+    print('Calculating it metrics '+model+' '+site)
+    n_lags = 9
+    nbins = 11
+    it_dict = {}
+       
+    #create an array of the prepped x and y variables
+    M = np.stack((x_ss,y_prepped), axis = 1)
+    #Mswap is for caclulating the TE from Y -> X, we don't really need to do that
+    #because DO doesn't affect solar radiation, but it is needed for function
+    Mswap = np.stack((y_prepped, x_ss), axis = 1)
+    #x_bounds and y_bounds are for removing outliers
+    x_bounds = it_functions.find_bounds(M[:,0], 0.1, 99.9)
+    y_bounds = it_functions.find_bounds(M[:,1], 0.1, 99.9)
+    M_x_bound = np.delete(M, np.where((M[:,0] < x_bounds[0]*1.1) | (M[:,0] > x_bounds[1]*1.1)), axis = 0)
+    M_xy_bound = np.delete(M_x_bound, np.where((M_x_bound[:,1] < y_bounds[0]*1.1) | (M_x_bound[:,1] > y_bounds[1]*1.1)), axis = 0)
+
+    #calc it metrics and store in the dictionary it_dict
+    it_dict = it_functions.calc_it_metrics(M_xy_bound, Mswap, n_lags, nbins, calc_swap = False, alpha = 0.05, ncores = 7)
```

Comment:
I wonder if it might be better to break this function up and return `it_dict` instead of `max_it`. That way you could look at the TE, MI and correlation across different time lags before the rest of the function choose the `max_it` time lag from the dictionary.

Reply:
Hm. I guess I don't understand what is in `it_dict` that we don't have in `max_it`. In line 162-163 we are storing the TE for all time steps 0-9 in `max_it` and in lines 168-169 we do the same for MI. Is that what you are referring to? 

If so, maybe `max_it` is actually just a bad variable name b/c it contains a lot more than the max TE and MI. 

If not, I think you are right in suggesting that we save more of the raw information somehow.

Reply:
Ok, so I think that `it_dict` has TE, MI, and correlation for all time lags (0-9) and `max_it` is a subset of `it_dict` where the maximum MI and TE values are chooses. That's how I wrote it and if I am reading the code right, I think that's what it's doing

Reply:
Ah ok, so I just got it running on my machine and I was mistaken, the `max_it` object does include the values of MI and TE for all time lags, it just identifies the maximum value for both modeled and observed and adds in the RMSE and makes the output dictionary a little easier to work with. So I think it's fine as is then, maybe adding a comment to explain what the `max_it` object contains might be helpful

Reply:
Sounds good. I add those here: https://github.com/USGS-R/drb-do-ml/pull/182/commits/24c4b9cfdaf59cf5f4062f1eba212a121630756c